ipa-core/src/protocol/hybrid/shuffle.rs

```diff
@@ -0,0 +1,159 @@
+use crate::{
+    error::Error,
+    ff::boolean_array::{BooleanArray, BA112, BA144, BA64},
+    protocol::{
+        context::Context,
+        ipa_prf::{
+            boolean_ops::{expand_shared_array_in_place, extract_from_shared_array},
+            shuffle::Shuffle,
+        },
+    },
+    report::hybrid::IndistinguishableHybridReport,
+    secret_sharing::{
+        replicated::{semi_honest::AdditiveShare, ReplicatedSecretSharing},
+        SharedValue,
+    },
+};
+
+/// Shuffles a Vec of IndistinguishableHybridReport
+/// # Errors
+/// Propogates errors from ctx.shuffle
+#[tracing::instrument(name = "shuffle_inputs", skip_all)]
+pub async fn shuffle_hybrid_inputs<C, BK, V>(
+    ctx: C,
+    input: Vec<IndistinguishableHybridReport<BK, V>>,
+) -> Result<Vec<IndistinguishableHybridReport<BK, V>>, Error>
+where
+    C: Context + Shuffle,
+    BK: BooleanArray,
+    V: BooleanArray,
+{
+    let shuffle_input: Vec<AdditiveShare<BA112>> = input
+        .into_iter()
+        .map(|item| hybrid_report_to_shuffle_input::<BA112, BK, V>(&item))
+        .collect::<Vec<_>>();
+
+    let shuffled = ctx.shuffle::<BA112, BA144, _>(shuffle_input).await?;
+
+    Ok(shuffled
+        .into_iter()
+        .map(|item| shuffled_to_hybrid_report(&item))
+        .collect::<Vec<_>>())
+}
+
+/// Converts `IndistinguishableHybridReport` into
+/// an `AdditiveShare` needed for shuffle protocol.
+pub fn hybrid_report_to_shuffle_input<YS, BK, V>(
```

Comment:
would it make more sense to implement these as `From` traits. 

Right now, `ctx.shuffle` has the trait bound 
```
I: IntoIterator<Item = AdditiveShare<S>> + Send,
```
but you could also consider updating that so that `Item` implements conversion into `AdditiveShare<S>`, rather than being an `AdditiveShare` directly. 

Updating this now will either mean updating existing IPA to work the same way (yuck!), so maybe we wait for now.